bindings/c/foundationdb/fdb_c.h

```diff
@@ -214,6 +214,37 @@ typedef struct readgranulecontext {
 	int granuleParallelism;
 } FDBReadBlobGranuleContext;
 
+typedef struct bgfilepointer {
+	const uint8_t* filename_ptr;
+	int filename_length;
+	int64_t file_offset;
+	int64_t file_length;
+	int64_t full_file_length;
+	/* TODO: encryption keys would go here */
+} FDBBGFilePointer;
+
+#pragma pack(push, 4)
+typedef struct bgmutation {
+	uint8_t type;
+	int64_t version;
+	const uint8_t* param1_ptr;
+	int param1_length;
+	const uint8_t* param2_ptr;
+	int param2_length;
+} FDBBGMutation;
+
+typedef struct bgfiledescription {
+	FDBKeyRange key_range;
+	fdb_bool_t snapshot_present;
+	FDBBGFilePointer snapshot_file_pointer;
+	int delta_file_count;
+	FDBBGFilePointer* delta_files;
+	int memory_mutation_count;
+	FDBBGMutation* memory_mutations;
+	/* TODO: tenant info would go here */
+} FDBBGFileDescription;
+#pragma pack(pop)
```

Comment:
I see we have a lot of structs that we're introducing as part of our C API. Historically we had only exposed opaque types and use functions to interact with them.

Do we have a well fleshed out plan for how to manage versioning and compatibility when working with these structs? We should have one in place prior to releasing them.

Reply:
Not particularly, no.
Given that these would be relatively limited use we had hoped to mostly ignore that problem, but if there's a good way to change this to support that better then I agree we should do it

Reply:
The main issue is that once we've released something, it becomes more difficult to change it in the future. For anything that is in the public C interface, we should have an idea what we would need to do to evolve the interface over time, though it doesn't necessarily have to be the same as what we did before.

For previous functionality that was all exposed through functions, we had a mechanism to control what versions of our functions would be used using the API version. When dealing with structs, we could theoretically have multiple definitions for different versions and/or a deprecation and extension story that makes it possible to continue using the same struct as it gets changed, but for either of those we should consider what implications that has for users of the interface. For what it's worth, I think some similar concerns were encountered recently with the mapped ranges API, so it's possible someone who worked on that might have some additional thoughts.